ujt/transformers.py

```diff
@@ -65,6 +95,101 @@ def apply_views(elements, tag_map, view_list):
+        end_time: The end time of the time range to compute aggregate status over.
+        sli_type: The SLI type of interest.
+
+    Returns:
+        A new list of elements with the change over time classes applied.
+    """
+
+    node_name_sli_map = defaultdict(list)
+    for sli in slis:
+        node_name_sli_map[sli.node_name].append(sli)
+
+    output_elements = []
+    for element in elements:
+        ujt_id = element["data"]["ujt_id"]
+        if ujt_id not in node_name_sli_map:
+            output_elements.append(element.copy())
+        else:
+            element_copy = element.copy()
+            (
+                before_composite_sli,
+                after_composite_sli,
+            ) = generate_before_after_composite_slis(
+                node_name_sli_map[ujt_id],
+                start_time,
+                end_time,
+            )
+            change_over_time_class = (
+                utils.get_change_over_time_class_from_composite_slis(
+                    before_composite_sli,
+                    after_composite_sli,
+                )
+            )
+            element_copy["classes"] += f" {change_over_time_class}"
+            output_elements.append(element_copy)
+
+    return output_elements
+
+
+def generate_before_after_composite_slis(
+    slis: List[SLI], start_time: dt.datetime, end_time: dt.datetime
+) -> Tuple[Optional[SLI], Optional[SLI]]:
+    """Generates two composite SLIs from the slis in each half of the specified time range.
+
+    The composite SLI's value is the average of the individual SLI values in the appropriate range.
+
+    Args:
+        slis: A list of SLIs within the tine range
+        start_time: The start of the time range
+        end_time: The end of the time range
+
+    Returns:
+        A tuple of two composite SLIs. It will never return (None, None)
+    """
+    if slis == []:
+        raise ValueError
+
+    mid_time = start_time + (end_time - start_time) / 2
+    slis_before, slis_after = [], []
+    for sli in slis:
+        if sli.timestamp.ToDatetime() < mid_time:
+            slis_before.append(sli)
+        else:
+            slis_after.append(sli)
+
+    # Compute the average value before and after min_time, and note the appropriate class
+    composite_slis: List[Optional[SLI]] = [None, None]
+    for idx, sli_sublist in enumerate([slis_before, slis_after]):
+        if sli_sublist != []:
+            composite_sli = SLI()
+            # copy the slo bound/target values from the input,
+            # these should be constant across all slis in the input list
+            composite_sli.CopyFrom(sli_sublist[0])
+            # compute the average value, this can be more sophisticated
+            composite_sli.sli_value = sum([sli.sli_value for sli in sli_sublist]) / len(
+                sli_sublist
+            )
+            compute_status.compute_sli_status(composite_sli)
+
+            composite_slis[idx] = composite_sli
```

Comment:
I choose to generate these composite SLIs in a iterative manner (rather than explicitly writing each case for before and after) to avoid duplicating code. This feels nicer and more extensible to me, but not sure if it's better to write it explicitly.